srv/src/main/java/my/bookshop/handlers/CatalogServiceHandler.java

```diff
@@ -2,56 +2,201 @@
+	public void beforeAddReview(AddReviewContext context) {
+		String reviewer = context.getUserInfo().getName();
+		String bookId = (String) analyzer.analyze(context.getCqn()).targetKeys().get(Books.ID);
+
+		Result result = db.run(Select.from(CatalogService_.REVIEWS)
+				.where(review -> review.book_ID().eq(bookId).and(review.reviewer().eq(reviewer))));
+
+		if (result.first().isPresent()) {
+			throw new ServiceException(ErrorStatuses.METHOD_NOT_ALLOWED,
+					"User not allowed to add more than one review for a given book").messageTarget(Reviews_.class,
+							r -> r.reviewer());
+		}
+	}
+
+	/**
+	 * Handles the review creation from the given context.
+	 *
+	 * @param context {@link ReviewContext}
+	 */
+	@On(entity = Books_.CDS_NAME)
+	public void onAddReview(AddReviewContext context) {
+		Integer rating = context.getRating();
+		String title = context.getTitle();
+		String text = context.getText();
+
+		String username = context.getUserInfo().getName();
+		String bookId = (String) analyzer.analyze(context.getCqn()).targetKeys().get(Books.ID);
+
+		cds.gen.reviewservice.Reviews review = cds.gen.reviewservice.Reviews.create();
+		review.setBookId(bookId);
+		review.setReviewer(username);
+		review.setRating(rating);
+		review.setTitle(title);
+		review.setText(text);
+
+		Result res = reviewService.run(Insert.into(ReviewService_.REVIEWS).entry(review));
+		cds.gen.reviewservice.Reviews inserted = res.single(cds.gen.reviewservice.Reviews.class);
+
+		messages.success(MessageKeys.REVIEW_ADDED);
+
+		context.setResult(Struct.access(inserted).as(Reviews.class));
+	}
+
+	/**
+	 * Recalculates and sets the book rating after a new review for the given book.
+	 *
+	 * @param context {@link ReviewContext}
+	 */
+	@After(entity = Books_.CDS_NAME)
+	public void afterAddReview(AddReviewContext context) {
+		bookRatingService.setBookRating(context.getResult().getBookId());
 	}
 
 	@After(event = CdsService.EVENT_READ)
 	public void discountBooks(Stream<Books> books) {
-		books.filter(b -> b.getTitle() != null).forEach(
-			b -> {
-				loadStockIfNotSet(b);
-				discountBooksWithMoreThan111Stock(b);
+		books.filter(b -> b.getTitle() != null).forEach(b -> {
+			loadStockIfNotSet(b);
+			discountBooksWithMoreThan111Stock(b);
+		});
+	}
+
+	@After
+	public void setIsReviewable(CdsReadEventContext context, List<Books> books) {
+
+		String user = context.getUserInfo().getName();
+		List<String> bookIds = books.stream().map(b -> b.getId()).collect(Collectors.toList());
+
+		Select<Reviews_> query = Select.from(CatalogService_.BOOKS, b -> b.filter(b.ID().in(bookIds)).reviews())
+				.where(r -> r.reviewer().eq(user));
+
+		Set<String> reviewedBooks = db.run(query).streamOf(Reviews.class).map(Reviews::getBookId)
+				.collect(Collectors.toSet());
+
+		for (Books book : books) {
+			if (reviewedBooks.contains(book.getId())) {
+				book.setIsReviewable(false);
 			}
-		);
+		}
+	}
+
+	@On
+	public void onSubmitOrder(SubmitOrderContext context) {
```

Comment:
Should this action also create an order for the user?

Reply:
This is only used in the vue app so far, so I decided to keep it simple (as in the nodejs sample).